### PR TITLE
Feature: Formatted JSON Parse Tree output

### PR DESCRIPTION
**Pull request by noob77777**
# Feature: Formatted JSON Parse Tree output

## Description
Added nlohmann/json dependency. Generate parse tree as formatted json. Ready for Treant integration.

## Remaining tasks

- [x] Everything

```
/**
   * Grammar:
   * S -> ABA
   * A -> a
   * B -> b
   * 
   * input: "aba"
   */

{
    "S": {
        "A": {
            "a": null
        },
        "B": {
            "b": null
        },
        "A_1": {
            "a": null
        }
    }
}
```

## Further work
Integrate with Input argument parser to take -o file.json as output path.

### CODE REVIEW

File: src/parser/parser.cpp [hunk 1]
```diff
@@ -103,4 +104,70 @@ void Parser::ParseNextStep() {
   }
 }
 
+void Parser::BuildParseTree() {
+  parse_tree_ = json::object({});
```
Comment (bisakhmondal) [inline src/parser/parser.cpp]: place this into constructor

File: src/parser/parser.cpp [hunk 2]
```diff
@@ -103,4 +104,70 @@ void Parser::ParseNextStep() {
...
+    // augment_entities to handle duplicates
+    std::unordered_map<std::string, int> symbol_count;
+    std::unordered_map<std::string, std::string> default_name;
+    for (auto &entity : entities) {
+      auto p_entity = entity;
+      if (symbol_count[entity]++ != 0) {
+        entity += "_" + std::to_string(symbol_count[entity] - 1);
+      }
+      default_name[entity] = p_entity;
+    }
```
Comment (bisakhmondal) [inline src/parser/parser.cpp]: will this ever work,  the if block?
each time entity is getting changed (taken through reference), resulting in a cascading effect.

**Review by bisakhmondal — COMMENTED** [review]
rest is okay

File: src/parser/parser.cpp (hunk 2 above)
Comment (TheSYNcoder) [inline src/parser/parser.cpp]: The line 
```
entity += "_" + std::to_string(symbol_count[entity] - 1);
```
changes the actual value of entity !

Can it be like
```
for ( auto entity : entities)
...
```

Comment (noob77777) [reply src/parser/parser.cpp]: intented

Comment (noob77777) [reply src/parser/parser.cpp]: intended

Comment (noob77777) [reply src/parser/parser.cpp]: if <A, A, C, A, B> this loop converts it to <A, A_1, C, A_2, B> in place

Comment (TheSYNcoder) [reply src/parser/parser.cpp]: Okay lgtm.

**Review by TheSYNcoder — DISMISSED** [review]

File: src/parser/parser.cpp (hunk 2 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: just swap the role of `p_entity` and `entity`. Then I guess we don't have to use keep track of the old symbol `default_name[entity] = p_entity;`.
Will it work?

File: src/parser/parser.cpp [hunk 3]
```diff
@@ -103,4 +104,74 @@ void Parser::ParseNextStep() {
...
+    json *parent_node = parent_node_stack.top();
+    // add entities to current parent node
+    for (const auto &entity : entities) {
+      if (std::find(terminals.begin(), terminals.end(), default_name[entity]) != terminals.end()) {
+        (*parent_node)[entity] = json();
+      } else {
+        (*parent_node)[entity] = json::object({});
+      }
+    }
+
```
Comment (bisakhmondal) [inline src/parser/parser.cpp]: also merge this into the previous loop

File: src/parser/parser.cpp (hunk 2 above)
Comment (TheSYNcoder) [reply src/parser/parser.cpp]: I think it is just changing the names and vouching for the old values.

File: src/parser/parser.cpp (hunk 3 above)
Comment (noob77777) [reply src/parser/parser.cpp:155]: this loop runs in reverse order

File: src/parser/parser.cpp [hunk 4]
```diff
@@ -103,4 +104,74 @@ void Parser::ParseNextStep() {
...
+    // update parent_node_stack
+    parent_node_stack.pop();
+    std::reverse(entities.begin(), entities.end());
+    for (const auto &entity : entities) {
```
Comment (bisakhmondal) [inline src/parser/parser.cpp]: use `rbegin iterator`. Unnecessary reversal!!

File: src/parser/parser.cpp (hunk 3 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: are you sure? ig it's not that loop

Comment (noob77777) [reply src/parser/parser.cpp]: ? :eyes:

Comment (bisakhmondal) [reply src/parser/parser.cpp]: Hey smart as $🤣
image

**Review by bisakhmondal — APPROVED** [review]
lgtm

File: src/parser/parser.cpp [hunk 5]
```diff
@@ -103,4 +104,70 @@ void Parser::ParseNextStep() {
...
+        (*parent_node)[entity] = json();
+      } else {
+        (*parent_node)[entity] = json::object({});
+      }
+    }
+
+    // update parent_node_stack
+    parent_node_stack.pop();
+    for (auto it = entities.rbegin(); it < entities.rend(); it++) {
+      if (std::find(non_terminals.begin(), non_terminals.end(), default_name[*it]) != non_terminals.end()) {
```
Comment (TheSYNcoder) [inline src/parser/parser.cpp:156]: Range based loop is preferred
```
for ( auto & x : std::vector<std::string>( entities.rbegin(), entities.rend()) {

}
```

Comment (bisakhmondal) [reply src/parser/parser.cpp:156]: NO

Comment (noob77777) [reply src/parser/parser.cpp:156]: copying...
same issue as last implementation

Comment (bisakhmondal) [reply src/parser/parser.cpp:156]: it creates another copy

Comment (TheSYNcoder) [reply src/parser/parser.cpp:156]: WTF

Comment (TheSYNcoder) [reply src/parser/parser.cpp:156]: Well ..

**Pull request merged**